File: LeetCode/2976-Minimum-Cost-to-Convert-String-I.cpp
// 2976. Minimum Cost to Convert String I

class Solution {
public:
    long long minimumCost(string source, string target, vector<char>& original, vector<char>& changed, vector<int>& cost) {
        int len = original.size();

        vector<vector<int>> adj(26, vector<int>(26, 1e9));
        for(int i = 0; i < 26; i++) adj[i][i] = 0;

        for(int x = 0; x < len; x++) {
            int i = original[x] - 'a';
            int j = changed[x] - 'a';
            adj[i][j] = min(adj[i][j], cost[x]);
        }

        // Floyd Warshall
        for(int via = 0; via < 26; via++) {
            for(int i = 0; i < 26; i++) {
                for(int j = 0; j < 26; j++) {
                    adj[i][j] = min(adj[i][j], adj[i][via] + adj[via][j]);
                }
            }
        }

        long long total_cost = 0;
        for(int x = 0; x< source.length(); x++) {
            int i = source[x] - 'a';
            int j = target[x] - 'a';
            if(adj[i][j] == 1e9) return -1;
            total_cost += adj[i][j];
        }

        return total_cost;
    }
};

// ==================================================================================

class Solution {
public:

    vector<long long> dijkstra(int start, vector<vector<pair<int,int>>>& adj) {

        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> pq;
        pq.push({0, start});

        vector<long long> dist(26, 1e15);
        dist[start] = 0;

        while(!pq.empty()) {
            int du = pq.top().first;
            int u = pq.top().second;
            pq.pop();
            for(auto it : adj[u]) {
                int v = it.first;
                int dv = it.second;
                if(du + dv < dist[v]) {
                    pq.push({du + dv, v});
                    dist[v] = du + dv;
                }
            }
        }

        return dist;
    }

    long long minimumCost(string source, string target, vector<char>& original, vector<char>& changed, vector<int>& cost) {
        vector<vector<pair<int,int>>> adj(26);
        int n = original.size();
        for(int i = 0; i < n; i++) {
            adj[original[i] - 'a'].push_back({changed[i] - 'a', cost[i]});
        }

        vector<vector<long long>> uTOv(26);
        for(int i = 0; i < 26; i++) {
            uTOv[i] = dijkstra(i, adj);
        }

        long long cst = 0, len = source.length();
        for(int i = 0; i < len; i++) {
            int u = source[i] - 'a', v = target[i] - 'a';
            if(uTOv[u][v] == 1e15) return -1;
            cst += uTOv[u][v];
        }

        return cst;
    }
};

File: LeetCode/2976-Minimum-Cost-to-Convert-String-I.java
// 2976. Minimum Cost to Convert String I

class Solution {
    public List<int[]>[] adj;
    public int INF = (int)1e9;

    public void getMinDist(int start, int[] dist) {
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[1] - b[1]);
        pq.add(new int[]{start, 0});
        dist[start] = 0;
        while (!pq.isEmpty()) {
            int[] curr = pq.poll();
            int u = curr[0];
            int du = curr[1];
            for (int[] edge : adj[u]) {
                int v = edge[0];
                int dv = edge[1];
                int cost = du + dv;
                if (cost < dist[v]) {
                    dist[v] = cost;
                    pq.add(new int[]{v, cost});
                }
            }
        }
    }

    public long minimumCost(String source, String target, char[] original, char[] changed, int[] cost) {
        adj = new ArrayList[26];
        for (int i = 0; i < 26; i++) adj[i] = new ArrayList<>();

        for (int i = 0; i < original.length; i++) {
            int u = original[i] - 'a';
            int v = changed[i] - 'a';
            int c = cost[i];
            adj[u].add(new int[]{v, c});
        }

        int[][] dist = new int[26][26];
        for (int i = 0; i < 26; i++) {
            Arrays.fill(dist[i], INF);
        }

        for (int i = 0; i < 26; i++) {
            getMinDist(i, dist[i]);
        }

        long ans = 0;
        for (int i = 0; i < source.length(); i++) {
            int u = source.charAt(i) - 'a';
            int v = target.charAt(i) - 'a';
            if (dist[u][v] == INF) return -1;
            ans += dist[u][v];
        }

        return ans;
    }
}
